services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala

```diff
@@ -1,29 +1,81 @@
 package cromwell.services.metadata.impl
 
-import akka.actor.{Actor, ActorLogging, Props}
+import akka.actor.{ActorLogging, LoggingFSM, Props}
 import cromwell.core.Dispatcher.ServiceDispatcher
 import cromwell.services.SingletonServicesStore
-import cromwell.services.metadata.MetadataService.{MetadataPutAcknowledgement, MetadataPutFailed, PutMetadataAction}
+import cromwell.services.metadata.MetadataEvent
+import cromwell.services.metadata.MetadataService.PutMetadataAction
+import cromwell.services.metadata.impl.WriteMetadataActor.WriteMetadataActorState
 
+import scala.concurrent.duration._
 import scala.util.{Failure, Success}
 
-object WriteMetadataActor {
-  def props() = Props(new WriteMetadataActor()).withDispatcher(ServiceDispatcher)
-}
-
-class WriteMetadataActor extends Actor with ActorLogging with MetadataDatabaseAccess with SingletonServicesStore {
+final case class WriteMetadataActor(batchRate: Int, flushRate: FiniteDuration)
+  extends LoggingFSM[WriteMetadataActorState, Vector[MetadataEvent]] with ActorLogging with
+  MetadataDatabaseAccess with SingletonServicesStore {
+  import WriteMetadataActor._
 
   implicit val ec = context.dispatcher
 
-  def receive = {
-    case action@PutMetadataAction(events) =>
-      val sndr = sender()
-      addMetadataEvents(events) onComplete {
-        case Success(_) => sndr ! MetadataPutAcknowledgement(action)
-        case Failure(t) =>
-          val msg = MetadataPutFailed(action, t)
-          log.error(t, "Sending {} failure message {}", sndr, msg)
-          sndr ! msg
+  override def preStart(): Unit = {
+    context.system.scheduler.schedule(0.seconds, flushRate, self, ScheduledFlushToDb)
+    super.preStart()
+  }
+
+  startWith(WaitingToWrite, EmptyVector)
+
+  when(WaitingToWrite) {
+    case Event(PutMetadataAction(events), curData) =>
+      val newData = curData ++ events.toVector
+      newData match {
+        case data if data.size > batchRate => goto(WritingToDb) using data
+        case e => stay using e
+      }
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic metadata flush to DB")
+      goto(WritingToDb) using curData
+  }
+
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay using curData
+    case Event(PutMetadataAction(events), curData) =>
+      val newData = curData ++ events.toVector
```

Comment:
Note to self: Cats provides a NonEmptyVector, investigate performance. ADT might be back in play